File: LeetCode-Cpp/TextJustification.cpp
class Solution {
public:
    vector<string> fullJustify(vector<string>& words, int maxWidth) {
         vector<string> result;
        vector<string> currentLine;
        int currentLength = 0;

        for (const string& word : words) {
            if (currentLength + word.length() + currentLine.size() > maxWidth) {
                result.push_back(justifyLine(currentLine, maxWidth, false));
                currentLine.clear();
                currentLength = 0;
            }

            currentLine.push_back(word);
            currentLength += word.length();
        }

        // Justify the last line (left-justified)
        result.push_back(justifyLine(currentLine, maxWidth, true));

        return result;
    }
    string justifyLine(const vector<string>& words, int maxWidth, bool isLastLine) {
    if (isLastLine || words.size() == 1) {
        string line = words[0];
        for (int i = 1; i < words.size(); ++i) {
            line += ' ' + words[i];
        }
        line += string(maxWidth - line.length(), ' ');
        return line;
    } else {
        int totalSpaces = maxWidth - accumulate(words.begin(), words.end(), 0, [](int sum, const string& word) { return sum + word.length(); });
        int numGaps = words.size() - 1;
        int spacesBetweenWords = totalSpaces / numGaps;
        int extraSpaces = totalSpaces % numGaps;

        string justifiedLine = words[0];
        for (int i = 1; i < words.size(); ++i) {
            int spaces = spacesBetweenWords + (i <= extraSpaces ? 1 : 0);
            justifiedLine += string(spaces, ' ') + words[i];
        }

        return justifiedLine;
    }
}
};

File: LeetCode-Python/TextJustification.py
class Solution:
    def fullJustify(self, words: List[str], maxWidth: int) -> List[str]:
        result = []
        current_line = []
        current_length = 0

        def justify_line(words, spaces, is_last_line):
            if is_last_line or len(words) == 1:
                return (' '.join(words)).ljust(maxWidth)
            else:
                extra_spaces = maxWidth - sum(len(word) for word in words)
                spaces_between_words = extra_spaces // (len(words) - 1)
                extra_spaces %= (len(words) - 1)

                justified_line = words[0]
                for i in range(1, len(words)):
                    spaces = spaces_between_words + (1 if i <= extra_spaces else 0)
                    justified_line += ' ' * spaces + words[i]

                return justified_line

        for word in words:
            if current_length + len(word) + len(current_line) > maxWidth:
                result.append(justify_line(current_line, maxWidth - current_length, False))
                current_line = []
                current_length = 0

            current_line.append(word)
            current_length += len(word)

        # Justify the last line (left-justified)
        result.append(justify_line(current_line, 1, True))

        return result


  
